Add DEFERRED datasets to Galaxy.

jmchilton opened the pull request:
- Improves options and display around discarded datasets.
- Adds new APIs for importing model stores.
- Adds support to model store layer for importing/exporting workflow invocations and APIs and tests to support this.
- Add new DEFERRED dataset state - with APIs for materializing them or the option to defer them on the worker node if using ``tool_evaluation_strategy=remote`` on the job destination.
- UI and API options for uploading deferred datasets individually or into collections.


## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

mvdbeek commented on lib/galaxy/model/__init__.py:
```diff
@@ -3151,6 +3158,7 @@ def values(self):
         states.OK,
         states.EMPTY,
         states.ERROR,
+        states.DIFFERED,
```
```suggestion
        states.DEFFERED,
```

jmchilton committed:
Add DEFERRED datasets to Galaxy

- Improves options and display around discarded datasets.
- Adds new APIs for importing model stores.
- Adds support to model store layer for importing/exporting workflow invocations and APIs and tests to support this.
- Add new DEFERRED dataset state - with APIs for materializing them or the option to defer them on the worker node if using tool_evaluation_strategy=remote on the job destination.
- UI and API options for uploading deferred datasets individually or into collections.

mvdbeek commented on lib/galaxy/model/deferred.py:
```diff
@@ -0,0 +1,203 @@
+
+    @abc.abstractmethod
+    def transient_paths_for(self, old_dataset: Dataset) -> TransientDatasetPaths:
+        """Map datset to transient paths for detached HDAs.
```
```suggestion
        """Map dataset to transient paths for detached HDAs.
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/history_contents.py:
```diff
@@ -293,6 +328,28 @@ def create(self, trans, history_id, payload, **kwd):
+        """
+        POST /api/histories/{history_id}/contents_from_store
+        Create history contents from model store.
+        Input can a tarfile created with build_objects script distributed
```
```suggestion
        Input can be a tarfile created with build_objects script distributed
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/history_contents.py:
```diff
@@ -293,6 +328,28 @@ def create(self, trans, history_id, payload, **kwd):
+        POST /api/histories/{history_id}/contents_from_store
+        Create history contents from model store.
+        Input can a tarfile created with build_objects script distributed
+        with galaxy-data, from an exported history with files stipped out,
```
```suggestion
        with galaxy-data, from an exported history with files stripped out,
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/short_term_storage.py:
```diff
@@ -0,0 +1,58 @@
+
+@router.cbv
+class FastAPIShortTermStorage:
+    # typing here is ideal, mypy is the issue xref https://github.com/python/mypy/issues/5374
```
```suggestion
    # typing here is not ideal, mypy is the issue xref https://github.com/python/mypy/issues/5374
```

mvdbeek replied:
I guess it'd still be nice to narrow the type ignore for future-us

mvdbeek commented on lib/galaxy/webapps/galaxy/services/base.py:
```diff
@@ -85,3 +115,93 @@ def get_authenticated_user(self, trans: ProvidesUserContext) -> User:
+    export_target: Any
+
+
+def model_store_stoarge_target(short_term_storage_allocator: ShortTermStorageAllocator, file_name: str, model_store_format: str) -> ShortTermStorageTarget:
```
```suggestion
def model_store_storage_target(short_term_storage_allocator: ShortTermStorageAllocator, file_name: str, model_store_format: str) -> ShortTermStorageTarget:
```

wm75 commented on lib/galaxy/datatypes/registry.py:
```diff
@@ -544,6 +557,28 @@ def load_datatype_sniffers(self, root, deactivate=False, handling_proprietary_da
+        for extension_segments in [3, 2, 1]:
+            if len(name_parts) < extension_segments + 1:
+                continue
+            possible_extensions.append(".".join(name_parts[-extension_segments:]))
```
```suggestion
    def get_datatype_from_filename(self, name, max_extension_parts=3):
        generic_datatype_instance = self.get_datatype_by_extension("data")
        if "." not in name:
            return generic_datatype_instance
        extension_parts = name.rsplit(".", max_extension_parts)[1:]
        possible_extensions = []
        for n, ext_part in enumerate(extension_parts):
            possible_extensions.append(".".join(extension_parts[n:]))
```

wm75 commented on lib/galaxy/datatypes/registry.py:
```diff
@@ -544,6 +557,28 @@ def load_datatype_sniffers(self, root, deactivate=False, handling_proprietary_da
+            possible_extensions.append(".".join(name_parts[-extension_segments:]))
+
+        infer_from = self.datatypes_by_suffix_inferences
+        for possible_extension in possible_extensions:
+            if possible_extension in infer_from:
+                return infer_from[possible_extension]
+
+        for possible_extension in possible_extensions:
+            if possible_extension in self.datatypes_by_extension:
+                return self.datatypes_by_extension[possible_extension]
```
Is this the intended order of lookups? Giving `infer_from` priority over `datatypes_by_extension`?

jmchilton replied:
In the case of inferring things from the filename - yes - I did intend to do that. The Galaxy "extension" doesn't really map well to actual file extensions even for really basic things like fastq and fasta files.

bgruening commented on lib/galaxy/config/schemas/config_schema.yml:
```diff
@@ -493,6 +493,34 @@ mapping:
+        type: int
+        required: false
+        desc: |
+          Default duration before short term storage files can be cleaned up.
```
same description as above

bgruening commented on lib/galaxy/exceptions/error_codes.json:
```diff
@@ -4,6 +4,11 @@
+   {
+      "name": "NO_CONTENT_GENERIC",
+      "code": 204001,
+      "message": "Galaxy has not content to associate with this request."
```
s/not/no/

bgruening commented on lib/galaxy/model/store/load_objects.py:
```diff
@@ -0,0 +1,76 @@
+See the corresponding galaxy-build-objects script for one possible way to
+create a model store to use with this script.
+
+This script "uploads" all datasets in a "discarded" state. To actually
```
discarded or deferred?

bgruening commented on lib/galaxy/webapps/galaxy/api/history_contents.py:
```diff
@@ -407,6 +412,31 @@ def show(
 
+    @router.get(
+        "/api/histories/{history_id}/contents/{type}s/{id}/prepare_store_download",
+        summary="TODO",
```
description missing

bgruening commented on lib/galaxy/webapps/galaxy/api/history_contents.py:
```diff
@@ -449,6 +479,26 @@ def download_dataset_collection(
 
+    @router.get(
+        "/api/histories/{history_id}/contents/dataset_collections/{id}/prepare_download",
+        summary="TODO",
```
missing

bgruening commented on lib/galaxy/webapps/galaxy/api/history_contents.py:
```diff
@@ -721,3 +771,42 @@ def contents_near(
+        # "/api/histories/{history_id}/contents/datasets/{id}/materialize",
+        #   id: EncodedDatabaseIdField = HistoryItemIDPathParam,
+        "/api/histories/{history_id}/materialize",
+        summary="Materialize a deferred dataset into real, usable dataset.",
```
The comment indicates that it is working for __one__ dataset, but the API looks like the entire history.

jmchilton replied:
I think I took pass at making this more clear. Let me know if it is still confusing.

jmchilton committed:
Rework galaxy.model.store.discover to not require unique paths for incoming files.

mvdbeek commented on lib/galaxy/jobs/handler.py:
```diff
@@ -596,7 +597,12 @@ def __filter_jobs_with_invalid_input_states(self, jobs):
                     or_(
                         model.Dataset.deleted == true(),
-                        model.Dataset.state != model.Dataset.states.OK,
+                        not_(
+                            or_(
+                                model.Dataset.state == model.Dataset.states.OK,
+                                model.Dataset.state == model.Dataset.states.DEFERRED,
+                            )
+                        ),
```
```suggestion
                        model.Dataset.state.not_in_((model.Dataset.states.OK, model.Dataset.states.DEFERRED)),
```
It might render to the same SQL, but reads a little easier ?

jmchilton replied:
If I'm recalling currently, in my experience this can render a much slower SQL. I'm not saying it won't optimize it - but I've seen past versions of SQL Alchemy implement this very directly and the result was slower. I will change it if you want though 😟. 

Update: I absolutely agree with you that your version reads more cleanly though.

mvdbeek commented on lib/galaxy/managers/context.py:
```diff
@@ -197,6 +198,10 @@ class ProvidesUserContext(ProvidesAppContext):
 
+    @property
+    def async_request_user(self) -> RequestUser:
+        return RequestUser(user_id=self.user.id)
```
Based on `get_current_user_roles` / `user_is_active` it looks like the user can be None, do we have to guard for that ?

jmchilton replied:
Done!

mvdbeek commented on lib/galaxy/managers/model_stores.py:
```diff
@@ -36,3 +70,179 @@ def setup_history_export_job(self, request: SetupHistoryExportJob):
+        tf.write(base64.b64decode(source_content))
+        tf.flush()
+        temp_dir = mkdtemp()
+        target_dir = os.path.abspath(CompressedFile(tf.name).extract(temp_dir))
```
```suggestion
        with NamedTemporaryFile("wb") as tf
            tf.write(base64.b64decode(source_content))
            tf.flush()
            temp_dir = mkdtemp()
            target_dir = os.path.abspath(CompressedFile(tf.name).extract(temp_dir))
```

mvdbeek commented on lib/galaxy/managers/model_stores.py:
```diff
@@ -36,3 +70,179 @@ def setup_history_export_job(self, request: SetupHistoryExportJob):
+        tf.write(base64.b64decode(source_content))
+        tf.flush()
+        temp_dir = mkdtemp()
+        target_dir = CompressedFile(tf.name).extract(temp_dir)
```
That's alsmost the same as the code in payload_to_source_uri, maybe a `extract_encoded_content_to_dir` utility function is a good idea ?

mvdbeek commented on lib/galaxy/managers/workflows.py:
```diff
@@ -385,6 +385,23 @@ def ensure_raw_description(self, dict_or_raw_description):
 
+    def read_workflow_from_path(self, app, user, path, allow_in_directory=None):
+        from galaxy.work.context import WorkRequestContext
+
```
Is the local import needed ?

jmchilton replied:
~We're doing it elsewhere in that file - I'll open a new PR against dev to see if there are any problem irrespective of this PR.~

Ignore that - it isn't like that upstream.

mvdbeek commented on lib/galaxy/model/__init__.py:
```diff
@@ -4335,6 +4372,10 @@ def __init__(
+        if not attached and object_session(self):
+            # add_object_to_object_session wasn't called, why do I need to do this?
+            # I guess a behavior that is changing with the next version of sqlalchemy?
+            object_session(self).expunge(self)
```
Is that still an issue after https://github.com/galaxyproject/galaxy/pull/13670?

jdavcs replied:
`add_object_to_object_session` is called in the `DatasetInstance` constructor (and before that call was added, the HDA object was added to the session implicitly on `self.dataset = dataset`)

jdavcs replied:
☝️ ...which means #13670 did not fix the implicit attachment issue?

jmchilton replied:
Trying a version of this without the hack in https://github.com/galaxyproject/galaxy/pull/13920/.

mvdbeek commented on lib/galaxy/model/deferred.py:
```diff
@@ -0,0 +1,282 @@
+        )
+        if require_metadata_regeneration:
+            if attached and self._sa_session:
+                # as of mid April 2022, we now get JSON encoding errors if this
```
Is there a test that fails ? This should be possible, https://github.com/galaxyproject/galaxy/commit/14dceed32b6a1b6f89147d1d532fd6a818007548 might be somewhat relevant ?

jmchilton replied:
I am sure some test failed or I would have never caught this. I couldn't point at the test though.

mvdbeek commented on lib/galaxy/model/deferred.py:
```diff
@@ -0,0 +1,282 @@
+        dataset_instance: Union[HistoryDatasetAssociation, LibraryDatasetDatasetAssociation],
+        target_history: Optional[History] = None,
+    ) -> HistoryDatasetAssociation:
+        """Create a new detached dataset instance from the supplied instance.
```
I'm having difficulty understanding the comment. Does detached refer to the sqlalchemy session or the object store ? It seems that based on whether or not you set the `attached` argument on the instance you may or may not be creating the dataset in the provided object store. If it refers to the sqlalchemy session then it also seems to depend (currently, but I think we can fix this)

jmchilton replied:
It is both - the detached instances are not written to an object store or the database. A "detached" (in the SQL Alchemy sense) HDA is created for tool evaluation that has a direct path encoded instead of relying on an object store. When we do the remote evaluation thing we want it to be detached in both senses right?

mvdbeek commented on lib/galaxy/model/deferred.py:
```diff
@@ -0,0 +1,282 @@
+                # isn't bound to the session before metadata generation.
+                self._sa_session.add(materialized_dataset_instance)
+            materialized_dataset_instance.init_meta()
+            materialized_dataset_instance.set_meta()
```
**Ignore the comment if this is only being called from within celery tasks.**

We shouldn't run this within handlers, this can take a long time and block all other threads.

jmchilton replied:
I'm imagining this running within the job script as part of your remote evaluation. evaluation.py is the main entry point here - though there is a Celery tasks to do this explicitly.

Maybe that isn't enough and you want to default to not doing evaluation.py in job handlers of deferred datasets. I think here we would want to start guarding everything behind some flags then. enable_beta_* at a level or two. I can do that if you'd like but I was hoping not to.

mvdbeek commented on lib/galaxy/model/store/load_objects.py:
```diff
@@ -0,0 +1,76 @@
+This script "uploads" all datasets in a "discarded" state. To actually
+load the datasets into a Galaxy instance - the underlying libraries need
+to be used directly and fed your Galaxy's database configuration and
+objectstore setup.
```
```suggestion
This script creates all datasets in a "discarded" state. To actually
load the datasets into a Galaxy instance's object store the underlying libraries need
to be used directly and fed your Galaxy's database configuration and
objectstore setup.
```
What libraries are you referring to ?

mvdbeek commented on lib/galaxy/schema/schema.py:
```diff
@@ -2786,6 +2796,27 @@ class PageSummaryBase(BaseModel):
+        description=(
+            "Depending on the `source` it can be:\n"
+            "- The encoded id from the library dataset\n"
+            "- The encoded id from the HDA\n"
```
```suggestion
            "- The encoded id of the source library dataset\n"
            "- The encoded id of the source HDA\n"
```
?

mvdbeek commented on lib/galaxy/schema/schema.py:
```diff
@@ -2786,6 +2796,27 @@ class PageSummaryBase(BaseModel):
+    source: DatasetSourceType = Field(
+        None,
+        title="Source",
+        description="The source of the content. Can be other history element to be copied or library elements.",
```
Should there be an enum of possible values here ?

jmchilton replied:
DatasetSourceType is an enum right? Or do you mean to include the possible values in the description string? I would think FastAPI would do that automatically?

mvdbeek replied:
Oh, right, yes, that should be sufficient.

mvdbeek commented on lib/galaxy/schema/tasks.py:
```diff
@@ -22,3 +31,63 @@ class GeneratePdfDownload(BaseModel):
+        description=(
+            "Depending on the `source` it can be:\n"
+            "- The unencoded id from the library dataset\n"
+            "- The unencoded id from the HDA\n"
```
```suggestion
            "- The decoded id of the source library dataset\n"
            "- The decoded id of the source HDA\n"
```

mvdbeek commented on lib/galaxy/tools/evaluation.py:
```diff
@@ -128,6 +139,27 @@ def validate_inputs(input, value, context, **kwargs):
+            if prefixed_name in undeferred_objects:
+                return undeferred_objects[prefixed_name]
+
+        visit_input_values(self.tool.inputs, incoming, replace_deferred)
```
Can we combine the three visit_input_values_calls into one ?

jmchilton replied:
I like the sequential way it run (validate, collect deferred datasets, undefer, replace) and I have no clue if validate_inputs needs to be run before the rest of it. It seems like if that is going to throw an exception - we should do that before we start materializing datasets.

I think the whole thing could take a pass for readability sake. Let me use comments and helpers to make that two separate operations a lot cleaner... and then see if you're okay with the way it reads? If you still want to merge at least the second two operations we can try to do that.

jmchilton replied:
This whole section of code became:

```python
        # Full parameter validation
        self._validate_incoming(incoming)

        # Restore input / output data lists
        inp_data, out_data, out_collections = job.io_dicts()

        # collect deferred datasets and collections.
        deferred_objects = self._deferred_objects(inp_data, incoming)

        # materialize deferred datasets
        materialized_objects = self._materialize_objects(deferred_objects, self.local_working_directory)

        # replace materialized objects back into tool input parameters
        self._replaced_deferred_objects(inp_data, incoming, materialized_objects)
```

Which feels very readable and clear to me. It feels cleaner and easier to work with in terms of future directions (e.g. caching of materialized datasets, parallelization) than trying to walk and materialize at the same time.

mvdbeek replied:
I feel like the walk is expensive (but maybe less so on a 2nd and 3rd round ?). If it turns out to be a problem we can refactor later.

mvdbeek commented on lib/galaxy/tools/evaluation.py:
```diff
@@ -196,6 +228,43 @@ def input():
         # Return the dictionary of parameters
         return param_dict
 
+    def _undeferred_objects(
```
Curious what the difference is between materialized and undeferred. If it's the same we should maybe consistently use one over the other.

jmchilton replied:
Hmmm... given my chaotic nature I'm sure I used them interchangeably. I guess all datasets that aren't deferred/discarded are materialized... and undeferred here implies the dataset was originally deferred and was materialized here.

I'll take a pass at readability here and see if I can make that better.

mvdbeek commented on lib/galaxy/tools/evaluation.py:
```diff
@@ -196,6 +228,43 @@ def input():
+                if value.state != model.Dataset.states.DEFERRED:
+                    continue
+
+                assert isinstance(value, (model.HistoryDatasetAssociation, model.LibraryDatasetDatasetAssociation))
```
```suggestion
                assert isinstance(value, model.DatasetInstance)
```
?

jmchilton replied:
The type checker didn't quite like that if I recall. The method explicitly takes in the DatasetInstance types it knows how to deal with right?

mvdbeek commented on lib/galaxy/tools/evaluation.py:
```diff
@@ -676,6 +747,8 @@ def build(self):
 class RemoteToolEvaluator(ToolEvaluator):
     """ToolEvaluator that skips unnecessary steps already executed during job setup."""
 
+    materialize_datasets = True
```
🥳

mvdbeek commented on lib/galaxy/web/framework/decorators.py:
```diff
@@ -338,7 +338,8 @@ def decorator(self, trans, *args, **kwargs):
                 raise validation_error_to_message_exception(e)
             if to_json:
-                rval = format_return_as_json(rval, jsonp_callback, pretty=trans.debug)
+                if not hasattr(rval, "read"):
```
What's this for ?

jmchilton replied:
Stray older stuff because I used message queues so heavily  - I'll remove it.

mvdbeek commented on lib/galaxy/webapps/galaxy/api/common.py:
```diff
@@ -51,25 +51,33 @@
     example="2015-01-29",
 )
 
+SeriializationFormatQueryParam: Optional[str] = Query(
```
```suggestion
SerializationFormatQueryParam: Optional[str] = Query(
```

jmchilton replied:
I think this is unused and left over also - I'm going to try to remove this wholesale. Sorry about that.

mvdbeek commented on lib/galaxy/webapps/galaxy/api/common.py:
```diff
@@ -51,25 +51,33 @@
     view: Optional[str] = SerializationViewQueryParam,
     keys: Optional[str] = SerializationKeysQueryParam,
     default_view: Optional[str] = SerializationDefaultViewQueryParam,
+    format: Optional[str] = SeriializationFormatQueryParam,
```
```suggestion
    format: Optional[str] = SerializationFormatQueryParam,
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/configuration.py:
```diff
@@ -70,7 +70,9 @@ def index(
-        return _index(self.configuration_manager, trans, view, keys)
+        index = _index(self.configuration_manager, trans, view, keys)
+        index["asgi_enabled"] = True
+        return index
```
```suggestion
        return _index(self.configuration_manager, trans, view, keys)
```
Should we just dispatch on the major version ? Starting from 22.05 every server is asgi_enabled

jmchilton replied:
I can't find this being used anyway. This was left over from the older style controllers.

mvdbeek commented on lib/galaxy/webapps/galaxy/api/histories.py:
```diff
@@ -262,6 +304,29 @@ def update(
+        self,
+        trans: ProvidesHistoryContext = DependsOnTrans,
+        serialization_params: SerializationParams = Depends(query_serialization_params),
+        payload: CreateHistoryFromStore = Body(...),
```
I don't know if `store_content_base64` is a good idea, for large store contents that could easily be cut off by the proxy.

jmchilton replied:
It is just used for testing and small things - making sure the ideas work - https://github.com/galaxyproject/galaxy/pull/13920 adds support for real URIs - HTTP/FTP/Galaxy Files/etc...

jmchilton replied:
That has been rebased into this PR and I've dropped the parameter for base64 imports. Hoping for something like #13947 to build these quick tests in the future.

mvdbeek commented on lib/galaxy/webapps/galaxy/api/history_contents.py:
```diff
@@ -714,6 +743,22 @@ def archive(
+        create_payload: CreateHistoryContentFromStore = Body(...),
+    ) -> List[AnyHistoryContentItem]:
+        """
+        Create history contents from model store.
+        Input can be a tarfile created with build_objects script distributed
+        with galaxy-data, from an exported history with files stripped out,
+        or hand-crafted JSON dictionary.
```
What's the advantage of this endpoint over using the history import ?

mvdbeek replied:
ok, that's a silly question after reading  more, sorry.

mvdbeek commented on lib/galaxy/webapps/galaxy/services/histories.py:
```diff
@@ -224,6 +244,42 @@ def create(
+        payload: CreateHistoryFromStore,
+    ) -> AsyncTaskResultSummary:  # TODO: serialized task somehow...
+        self._ensure_can_create_history(trans)
+        from galaxy.managers.model_stores import payload_to_source_uri
```
This should go to the top

jmchilton replied:
The comment about wanting to move this to model store was so I could feel more confident about that I think. Thanks for catching this. I'll see if I can refactor that update the import.

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -89,7 +103,23 @@
 
-class WorkflowsAPIController(BaseGalaxyAPIController, UsesStoredWorkflowMixin, UsesAnnotations, SharableMixin):
+class CreateInvocationFromStore(Model):
+    store_content_base64: Optional[str]
```
I haven't understood what `store_content_base64` offers. Is this so we don't have to support proper file uploads ?

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -89,7 +103,23 @@
+    store_dict: Optional[Dict[str, Any]]
+    history_id: Optional[str]
+
+    class Config:
+        extra = Extra.allow
```
What extra things are you expecting here ? I think it's good if we can avoid this so API consumers know when they're submitting invalid payloads.

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -89,7 +103,23 @@
-class WorkflowsAPIController(BaseGalaxyAPIController, UsesStoredWorkflowMixin, UsesAnnotations, SharableMixin):
+class CreateInvocationFromStore(Model):
+    store_content_base64: Optional[str]
+    store_dict: Optional[Dict[str, Any]]
```
It would be fantastic if we could model this at least partially

jmchilton replied:
This is meant to import stuff exported by Galaxy. It isn't meant to be assembled by developers but... it still I guess a schema would be helpful.

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -887,11 +918,37 @@ def index_invocations(self, trans: GalaxyWebTransaction, **kwd):
+
+        Create invocation(s) from a supplied model store.
+
+        Input can a tarfile created with build_objects script distributed
```
```suggestion
        Input can be a tarfile created with build_objects script distributed
```

mvdbeek replied:
Is the tarfile input the base64-encoded bytes of the tarfile ?

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -887,11 +918,37 @@ def index_invocations(self, trans: GalaxyWebTransaction, **kwd):
-        GET /api/workflows/{workflow_id}/invocations/{invocation_id}
-        GET /api/invocations/{invocation_id}
+        GET /api/workflows/{workflow_id}/invocations/{invocation_id}{.format}
+        GET /api/invocations/{invocation_id}{.format}
```
I think the format should probably be controlled by the accept header ?

jmchilton replied:
Again more leftover stuff - sorry. I'll clean this up.

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -61,17 +66,26 @@
+    async_task_summary,
+    ConsumesModelStores,
+    ensure_celery_tasks_enabled,
+    model_store_stoarge_target,
```
```suggestion
    model_store_storage_target,
```

mvdbeek replied:
(If you're coding on vscode, I find https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker quite useful).

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -918,13 +975,53 @@ def show_invocation(self, trans: GalaxyWebTransaction, invocation_id, **kwd):
+            invocation_name = f"Invocation of {workflow_invocation.workflow.stored_workflow.name} at {workflow_invocation.create_time.isoformat()}"
+        except AttributeError:
+            invocation_name = f"Invocation of workflow at {workflow_invocation.create_time.isoformat()}"
+        short_term_storage_target = model_store_stoarge_target(
```
```suggestion
        short_term_storage_target = model_store_storage_target(
```

mvdbeek commented on lib/galaxy/webapps/galaxy/services/base.py:
```diff
@@ -102,6 +115,53 @@ def get_authenticated_user(self, trans: ProvidesUserContext) -> User:
+    export_target: Any
+
+
+def model_store_stoarge_target(
```
```suggestion
def model_store_storage_target(
```

mvdbeek commented on lib/galaxy/webapps/galaxy/services/histories.py:
```diff
@@ -48,18 +55,29 @@
+from galaxy.webapps.galaxy.services.base import (
+    async_task_summary,
+    ConsumesModelStores,
+    model_store_stoarge_target,
```
```suggestion
    model_store_storage_target,
```

mvdbeek commented on lib/galaxy/webapps/galaxy/services/histories.py:
```diff
@@ -259,7 +315,41 @@ def show(
+        include_deleted: bool = False,
+    ) -> AsyncFile:
+        history = self.manager.get_accessible(self.decode_id(history_id), trans.user, current_history=trans.history)
+        short_term_storage_target = model_store_stoarge_target(
```
```suggestion
        short_term_storage_target = model_store_storage_target(
```

mvdbeek commented on lib/galaxy/webapps/galaxy/services/history_contents.py:
```diff
@@ -81,18 +87,26 @@
     async_task_summary,
+    ConsumesModelStores,
     ensure_celery_tasks_enabled,
+    model_store_stoarge_target,
```
```suggestion
    model_store_storage_target,
```

mvdbeek commented on lib/galaxy/webapps/galaxy/services/history_contents.py:
```diff
@@ -287,6 +347,40 @@ def show(
+            content_name = dataset_collection_instance.name
+        else:
+            raise exceptions.UnknownContentsType(f"Unknown contents type: {contents_type}")
+        short_term_storage_target = model_store_stoarge_target(
```
```suggestion
        short_term_storage_target = model_store_storage_target(
```

mvdbeek reviewed (commented):
This is all super exciting!

bgruening commented on lib/galaxy_test/base/populators.py:
```diff
@@ -1224,6 +1381,32 @@ def get_invocation(self, invocation_id, step_details=False):
+        create_response = self.create_invocation_from_store_raw(
+            history_id, store_dict=store_dict, store_path=store_path
+        )
+        print(create_response.content)
```
```suggestion
```

bgruening commented on lib/galaxy_test/base/populators.py:
```diff
@@ -1903,6 +2086,19 @@ def new_private_library(self, name):
+    ) -> List[Dict[str, Any]]:
+        payload = _store_payload(store_dict=store_dict, store_path=store_path)
+        create_response = self.create_from_store_raw(payload)
+        print(create_response.content)
```
```suggestion
```

bgruening commented on test/unit/data/test_dataset_materialization.py:
```diff
@@ -0,0 +1,353 @@
+
+
+def _assert_2_bed_metadata(hda: HistoryDatasetAssociation) -> None:
+    # metadata_dbkey: ?
```
should this comment be removed?

jmchilton replied:
Done.

bgruening commented on test/unit/data/test_dataset_materialization.py:
```diff
@@ -0,0 +1,353 @@
+
+
+def _ensure_relations_attached_and_expunge(deferred_hda: HistoryDatasetAssociation, fixture_context) -> None:
+    # make sure everything needed is in session (soures, hashes, and metadata)...
```
```suggestion
    # make sure everything needed is in session (sources, hashes, and metadata)...
```

jmchilton committed:
Task status API for celery.

jdavcs commented on lib/galaxy/model/migrations/alembic/versions_gxy/6a67bf27e6a6_deferred_data_tables.py:
```diff
@@ -0,0 +1,28 @@
+
+def downgrade():
+    op.drop_column("history_dataset_association", "metadata_deferred")
+    op.drop_column("library_dataset_dataset_association", "metadata_deferred")
```
Let's use the `drop_column` helper from `migrations.util` like we already do in the previous revision file (to work around sqlite limitations - see #13541)

```suggestion
    drop_column("history_dataset_association", "metadata_deferred")
    drop_column("library_dataset_dataset_association", "metadata_deferred")
```

jdavcs commented on lib/galaxy/model/migrations/alembic/versions_gxy/6a67bf27e6a6_deferred_data_tables.py:
```diff
@@ -0,0 +1,28 @@
+    Boolean,
+    Column,
+)
+
```
```suggestion

from galaxy.model.migrations.util import drop_column
```
```
